ostd/src/sync/spin.rs

```diff
@@ -16,65 +17,87 @@ use crate::{
 };
 
 /// A spin lock.
-pub struct SpinLock<T: ?Sized> {
+#[repr(transparent)]
+pub struct SpinLock<T: ?Sized, G = PreemptDisabled> {
+    phantom: PhantomData<G>,
+    /// Only the last field of a struct may have a dynamically sized type
+    /// That's why SpinLockInner is put in the last field
+    inner: SpinLockInner<T>,
+}
+
+struct SpinLockInner<T: ?Sized> {
     lock: AtomicBool,
     val: UnsafeCell<T>,
 }
 
-impl<T> SpinLock<T> {
+pub trait Guardian {
+    type Guard;
+
+    /// Return a guard object
+    fn guard() -> Self::Guard;
+}
```

Comment:
Public items should be well-documented. Try writing something like
```rust
/// A guardian that denotes the guard behavior for holding the spin lock.
pub trait Guardian {
    /// The guard type.
    type Guard;

    /// Creates a new guard.
    fn guard() -> Self::Guard;
}

/// A guardian that disables preemption while holding the spin lock.
pub struct PreemptDisabled;

/// A guardian that disables IRQs while holding the spin lock.
pub struct LocalIrqDisabled;
```

The documentation should be a well-structured sentence (e.g., capitalize the first latter and have the period) instead of something like "for lock disabling preempt". Take a look at the [official Rust guidelines](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components) on writing documentation to get a feel for how to write it.

Reply:
Actually, the type `Guard` and the method `guard()` in the trait `Guardian` are also public items, since members in a trait are always `pub`.

We should expect the OSTD to pass the [`missing_doc`](https://doc.rust-lang.org/rustdoc/lints.html#missing_docs) lint (although we don't currently enforce this in CI), so they should also have some documentations. Just something like the example given in https://github.com/asterinas/asterinas/pull/1149/files#r1715368872 should be enough:
```rust
    /// The guard type.
    type Guard;

    /// Creates a new guard.
    fn guard() -> Self::Guard;
```